libs/s25main/network/GameServer.cpp

```diff
@@ -430,13 +434,112 @@ void GameServer::Stop()
     LOG.write("server state changed to stop\n");
 }
 
+// Check if there are players that have not been assigned a team but only a
+// range. Those players are assigned a team now while we try to balanace the
+// number of players per team. Returns true iff players have been assigned.
+static bool adjustSmartTeams(std::vector<JoinPlayerInfo>& playerInfos)
+{
+    std::set<unsigned> nPotentialPlayers[4];
+    unsigned nPlayers[4] = {0, 0, 0, 0};
+    unsigned unassignedPlayers = 0;
+
+    // First collect fixed players and potential ones.
+    for(unsigned player = 0; player < playerInfos.size(); ++player)
+    {
+        JoinPlayerInfo& playerInfo = playerInfos[player];
+        switch(playerInfo.team)
+        {
+            case TM_TEAM1: ++nPlayers[0]; break;
+            case TM_TEAM2: ++nPlayers[1]; break;
+            case TM_TEAM3: ++nPlayers[2]; break;
+            case TM_TEAM4: ++nPlayers[3]; break;
+            case TM_TEAM_1_TO_2:
+                ++unassignedPlayers;
+                nPotentialPlayers[0].insert(player);
+                nPotentialPlayers[1].insert(player);
+                break;
+            case TM_TEAM_1_TO_3:
+                ++unassignedPlayers;
+                nPotentialPlayers[0].insert(player);
+                nPotentialPlayers[1].insert(player);
+                nPotentialPlayers[2].insert(player);
+                break;
+            case TM_TEAM_1_TO_4:
+                ++unassignedPlayers;
+                nPotentialPlayers[0].insert(player);
+                nPotentialPlayers[1].insert(player);
+                nPotentialPlayers[2].insert(player);
+                nPotentialPlayers[3].insert(player);
+                break;
+            case TM_NOTEAM: break;
+            default: RTTR_Assert_Msg(false, "Unexpected team!");
+        }
+    }
+    // Check for unassigned players first.
+    if(unassignedPlayers == 0)
+        return false;
+
+    RTTR_Assert_Msg(!nPotentialPlayers[0].empty() || !nPotentialPlayers[1].empty() || !nPotentialPlayers[2].empty()
+                      || !nPotentialPlayers[3].empty(),
+                    "Expected at least one team!");
+
+    while(unassignedPlayers)
+    {
+        // Set of potential teams for the next player.
+        std::vector<int> teamsForNextPlayer;
+
+        // Determine the minimal team size for teams that can take an unassigned player.
+        unsigned minNextTeamSize = -1;
+        for(int team = 0; team < 4; ++team)
+        {
+            // Check if we can add a player to this team at all.
+            if(!nPotentialPlayers[team].empty())
+              minNextTeamSize = std::min(minNextTeamSize, nPlayers[team]);
+        }
+        RTTR_Assert_Msg(minNextTeamSize != -1, "Expected to find team for an unassigned player!");
```

Comment:
This can be removed if you initialize `minNextTeamSize=0`

Reply:
But that would simply hide a broken invariant, right? The point was to ensure we do not end up here without having potential players to distribute, though, I can obviously remove the asserts if that is preferred.